### PR TITLE
support configure fitness through config files

### PR DESCRIPTION
**Pull request by LeDuySon**
## Description of the PR
Support to customize the fitness metrics through configuration yaml files

## What is fitness metrics
The fitness function calculates a weighted sum of four metrics:
`
fitness = (Precision * w[0]) + (Recall * w[1]) + (mAP@0.5 * w[2]) + (mAP@0.5:0.95 * w[3])
`
Where `w` is the `fitness_weight` list: `[P, R, mAP@0.5, mAP@0.5:0.95]`

So if you want to customize the target metrics to recall, you can set it:
`
fitness_weight: [0.0, 0.9, 0.1, 0.0] 
`
The recall in this case will have 90% contribution to final metrics

## Important note about evaluation metrics (fitness)
Fitness metric not directly affected the training process of ultralytic, ultralytic optimization on the loss function only. 

What fitness metric affect:

1. Best Model Selection: Only models with the highest weighted fitness score get saved as best.pt
2. Early Stopping: Training stops only when weighted fitness stops improving for patience epochs
3. Hyperparameter Tuning: The tuner searches for hyperparameters that maximize weighted fitness

## Summary by CodeRabbit

* **New Features**
  * Added support for customizable fitness metric weighting during training and evaluation, enabling users to adjust the importance of Precision, Recall, mAP@0.5, and mAP@0.5:0.95.
  * Introduced new minimal COCO and COCO-pose dataset configuration files for quick testing and development.
  * Provided a sample training script for running object detection and pose estimation tasks with the new configurations.

* **Bug Fixes**
  * Ensured consistent propagation and usage of fitness weight settings across detection, segmentation, pose, and OBB metric evaluations.

* **Tests**
  * Added comprehensive tests to verify correct handling and application of custom fitness weights in metric calculations and optimization strategies.

### CODE REVIEW

File: ultralytics/cfg/default.yaml [hunk 1]
```diff
@@ -37,6 +37,7 @@ fraction: 1.0 # (float) dataset fraction to train on (default is 1.0, all images
 profile: False # (bool) profile ONNX and TensorRT speeds during training for loggers
 freeze: # (int | list, optional) freeze first n layers, or freeze list of layer indices during training
 multi_scale: False # (bool) Whether to use multiscale during training
+fitness_weight: [0.0, 0.0, 0.1, 0.9] # (list[float]) fitness weights for [P, R, mAP@0.5, mAP@0.5:0.95], can be overridden for custom optimization targets
```
Comment (Kag-Bar) [inline ultralytics/cfg/default.yaml:40]: Are those the original default values, or are those just some you provided?

File: ultralytics/models/yolo/detect/val.py [hunk 2]
```diff
@@ -98,6 +98,10 @@ def init_metrics(self, model: torch.nn.Module) -> None:
         self.seen = 0
         self.jdict = []
         self.metrics.names = self.names
+        # Update metrics with fitness_weight from config if needed
+        if not hasattr(self.metrics.box, 'fitness_weight') or self.metrics.box.fitness_weight == [0.0, 0.0, 0.1, 0.9]:
```
Comment (Kag-Bar) [inline ultralytics/models/yolo/detect/val.py:102]: I am not sure i get the if statement logic -if the  fitness_weight  is not listed as the box attributes **or** it is defined exactly to [0.0, 0.0, 0.1, 0.9] we should defined it to be something else?

I dont follow on this one, please explain

File: scripts/train_model.py
Comment (Kag-Bar) [inline scripts/train_model.py:1]: is this a necessary script?

File: configs/coco8_pose.yaml
Comment (Kag-Bar) [inline configs/coco8_pose.yaml:1]: is this a necessary script? maybe keep it under "example" or "playground"?

File: configs/coco8_pose.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
+path: coco8-pose # dataset root dir
+train: images/train # train images (relative to 'path') 4 images
+val: images/val # val images (relative to 'path') 4 images
+test: # test images (optional)
+
+# Keypoints
+kpt_shape: [17, 3] # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)
+flip_idx: [0, 2, 1, 4, 3, 6, 5, 8, 7, 10, 9, 12, 11, 14, 13, 16, 15]
```
Comment (Kag-Bar) [inline configs/coco8_pose.yaml:19]: what is the flip idx?

File: configs/coco8.yaml
Comment (Kag-Bar) [inline configs/coco8.yaml:1]: is this a necessary script?

File: ultralytics/utils/metrics.py [hunk 4]
```diff
@@ -795,14 +796,15 @@ class Metric(SimpleClass):
...
-    def __init__(self) -> None:
+    def __init__(self, fitness_weight=None) -> None:
         """Initialize a Metric instance for computing evaluation metrics for the YOLOv8 model."""
         self.p = []  # (nc, )
         self.r = []  # (nc, )
         self.f1 = []  # (nc, )
         self.all_ap = []  # (nc, 10)
         self.ap_class_index = []  # (nc, )
         self.nc = 0
+        self.fitness_weight = fitness_weight or [0.0, 0.0, 0.1, 0.9]  # default weights
```
Comment (Kag-Bar) [inline ultralytics/utils/metrics.py:807]: why are [0.0, 0.0, 0.1, 0.9] the default values? what is the original values, if even used? 
I want to make sure we scale down to the original method of choosing the best model if those values are not provided

File: ultralytics/utils/metrics.py [hunk 5]
```diff
@@ -892,7 +894,10 @@ def maps(self) -> np.ndarray:
 
     def fitness(self) -> float:
         """Return model fitness as a weighted combination of metrics."""
-        w = [0.0, 0.0, 0.1, 0.9]  # weights for [P, R, mAP@0.5, mAP@0.5:0.95]
+        # weights for [P, R, mAP@0.5, mAP@0.5:0.95]
+        w = self.fitness_weight 
+        
+        print(f"*****************Fitness weight: {w}******************************")
         return (np.nan_to_num(np.array(self.mean_results())) * w).sum()
```
[the remark this answers is not in the document]
Comment (Kag-Bar) [reply ultralytics/utils/metrics.py]: agree with this comment

**Review by Kag-Bar — CHANGES_REQUESTED** [review]
we got to make sure we handle the default case as was used before introducing the metric choosing

File: ultralytics/cfg/default.yaml (hunk 1 above)
Comment (LeDuySon) [reply ultralytics/cfg/default.yaml:40]: @Kag-Bar  The default value

File: ultralytics/models/yolo/detect/val.py (hunk 2 above)
Comment (LeDuySon) [reply ultralytics/models/yolo/detect/val.py:102]: @Kag-Bar You can interpret it as if the fitness weight is not set or it has default value, we will try to read from config file to see whether or not we can update it

File: ultralytics/utils/metrics.py (hunk 4 above)
Comment (LeDuySon) [reply ultralytics/utils/metrics.py:807]: [0.0, 0.0, 0.1, 0.9] is the default values provided by ultralytics

File: configs/coco8_pose.yaml (hunk 3 above)
Comment (LeDuySon) [reply configs/coco8_pose.yaml:19]: I'm not sure, this config file i copy from ultralytic datasets just to test the training

File: configs/coco8_pose.yaml
Comment (LeDuySon) [reply configs/coco8_pose.yaml:1]: Yess will move it

File: configs/coco8.yaml
Comment (LeDuySon) [reply configs/coco8.yaml:1]: Just for testing, you can ignore it

File: ultralytics/models/yolo/detect/val.py (hunk 2 above)
Comment (Kag-Bar) [reply ultralytics/models/yolo/detect/val.py:102]: Got it
